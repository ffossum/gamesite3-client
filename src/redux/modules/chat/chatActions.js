/* @flow */
export const JOIN_CHANNEL = 'chat/join channel';
export const SEND_MESSAGE = 'chat/send message';
export const RECEIVE_MESSAGE = 'chat/receive message';

export type ChannelName = string;
export type UserMessage = {
  uid: string,
  txt: string,
};

type JoinChannelAction = {
  type: 'chat/join channel',
  payload: string,
};
export function joinChannel(channelName: ChannelName): JoinChannelAction {
  return {
    type: JOIN_CHANNEL,
    payload: channelName,
  };
}

type SendMessageAction = {
  type: 'chat/send message',
  payload: {
    ch: ChannelName,
    msg: UserMessage,
  },
};
export function sendMessage(
  ch: ChannelName,
  msg: UserMessage
): SendMessageAction {
  return {
    type: SEND_MESSAGE,
    payload: {
      ch,
      msg,
    },
  };
}

type ReceiveMessageAction = {
  type: 'chat/receive message',
  payload: {
    ch: ChannelName,
    msg: UserMessage,
  },
};
export function receiveMessage(
  ch: ChannelName,
  msg: UserMessage
): ReceiveMessageAction {
  return {
    type: RECEIVE_MESSAGE,
    payload: {
      ch,
      msg,
    },
  };
}

export type ChatAction =
  | JoinChannelAction
  | SendMessageAction
  | ReceiveMessageAction;
