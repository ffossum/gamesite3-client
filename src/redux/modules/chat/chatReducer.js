/* @flow */
import type { Action } from '../../actions';
import type { ChannelState } from './channelReducer';
import type { ChannelName } from './chatActions';

import { SEND_MESSAGE, RECEIVE_MESSAGE } from './chatActions';
import channelReducer from './channelReducer';

export type ChatState = {
  [ChannelName]: ChannelState,
};
const initialState: ChatState = {};

export default function chatReducer(
  state: ChatState = initialState,
  action: Action
) {
  switch (action.type) {
    case RECEIVE_MESSAGE:
    case SEND_MESSAGE: {
      const { ch } = action.payload;
      return {
        ...state,
        [ch]: channelReducer(state[ch], action),
      };
    }
    default:
      return state;
  }
}
