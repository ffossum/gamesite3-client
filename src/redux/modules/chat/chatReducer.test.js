/* @flow */
/* eslint-env jest */
import chatReducer from './chatReducer';
import { receiveMessage } from './chatActions';

describe('chat reducer', () => {
  const initialState = chatReducer(undefined, { type: '@@INIT' });

  test('initial state', () => {
    expect(initialState).toMatchSnapshot();
  });

  test('adds received message to channel', () => {
    const action = receiveMessage('mainchat', { uid: 'userid', txt: 'hello' });
    const state = chatReducer(initialState, action);

    expect(state.mainchat && state.mainchat.messages).toHaveLength(1);
    expect(state.mainchat && state.mainchat.messages[0]).toEqual({
      uid: 'userid',
      txt: 'hello',
    });
  });
});
