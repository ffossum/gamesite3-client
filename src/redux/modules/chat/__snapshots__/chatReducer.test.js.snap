// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat reducer initial state 1`] = `Object {}`;
